Dockerfile

```diff
@@ -0,0 +1,33 @@
+FROM haskell:8.6 as build
+WORKDIR /build
+RUN cabal new-update
+
+# Build our upstream dependencies after copying in only enough to tell cabal
+# what they are.  This will make these layers cache better even as we change the
+# code of semantic itself.
+COPY semantic.cabal /build/
+COPY cabal.project /build/
+COPY semantic-core/semantic-core.cabal /build/semantic-core/
+COPY vendor /build/vendor
+RUN cabal new-build --only-dependencies
+
+# Once the dependencies are built, copy in the rest of the code and compile
+# semantic itself.
+COPY . /build
+RUN cabal new-build semantic:exe:semantic
+
+# A fake `install` target until we can get `cabal new-install` to work
+RUN cp $(find dist-newstyle -name semantic -type f -perm -u=x) /usr/local/bin/semantic
+
+# Create a fresh image containing only the compiled CLI program, so that the
+# image isn't bulked up by all of the extra build state.
+FROM debian:stretch
```

Comment:
Might do `debian:stable-slim` or `debian:stretch-slim` here.

Reply:
Done.  (Used `stretch-slim` instead of `stable-slim` since `haskell:8.6` is based off of `stretch`.)